lib/ipmi_picmg.c

```diff
@@ -1373,59 +1375,97 @@ ipmi_picmg_get_led_state(struct ipmi_intf * intf, char ** argv)
 		return -1;
 	}
 
-	printf("LED states:						  %x	", rsp->data[1] );
-
-	if (!(rsp->data[1] & 0x1)) {
-		printf("[NO LOCAL CONTROL]\n");
-		return 0;
+	if (rsp->data_len != 5 && rsp->data_len != 8 && rsp->data_len != 9) {
+		lprintf(LOG_ERR, "LED get state failed with invalid response data length");
+		return -1;
 	}
 
-	printf("[LOCAL CONTROL");
+	printf("LED states:			  %x	[", rsp->data[1] );
 
-	if (rsp->data[1] & 0x2) {
-		printf("|OVERRIDE");
+	if (rsp->data[1] & 0x8) {
```

Comment:
While you're at it, could you please replace all the magic (at least in the affected code) with some meaningful macros/enums ? Please do that in a separate commit (first commit to replace magic, second one to actually fix the issue). Can you please do that?

Reply:
Since I'm new to Git, I'm not sure if I understand you correctly. When you say separate commits, do you mean two pulled branches? Should I create a separated issue/branch to create magic/macros?

Reply:
No, I mean a single branch (this one, `bugfix/297-fix-picmg-led-get`) with two separate commits in it on top of ipmitool/master branch. Git is different from subversion and cvs. In those systems you only have one operation, that is 'commit'. In git you have a 'commit' operation that commits an annotated chunk of your changes into your working copy, and then you have a 'push' operation that puts all the changes you have in your working copy into the remote repository (to github). That is what I mean. You replace the magic numbers, then do `git add -u; git commit`, then you fix the issue and again do `git add -u; git commit`. Then you do a single `git push -f` to replace the branch with your new set of changes. Then I can see during the review where you only did cosmetics and where you've actually changed the code.

I hope it's more clear now.

Thanks.

Reply:
Thank you for the detailed instructions.

However, since I've pushed the changes without magic numbers replaced, can I replace the magic numbers as the second step, and then do "git add -u; git commit; git push -f"?

Or should I delete and recreate the branch?

Reply:
By the way, could you please point me a file as an example of the coding convention for the magic numbers I should follow?

Reply:
https://www.burntfen.com/2015-10-30/how-to-amend-a-commit-on-a-github-pull-request

I tend to do Option 2, using `git commit --amend` then `git push --force`

Reply:
@PengWang-SMARTM, no, please do not delete the branch, that will ruin the connection to this PR.
If I want to reorder commits I usually do "interactive rebase".
That is, for this case I would do the following:

1. You did your code changes on branch `bugfix/297-fix-picmg-led-get`
2. Now you create a new temporary local branch for magic cleanup:
    `git checkout -b tmpbranch origin/master` (that is, the new branch doesn't contain your code changes)
3. You do you magic cleanup and commit it to `tmpbranch`:
4. You switch to  `bugfix/297-fix-picmg-led-get`:
    `git checkout
4. You rebase your temporary branch interactively to your `bugfix/297-fix-picmg-led-get`.
    It will show you two commits

Reply:
@PengWang-SMARTM, no, please do not delete the branch, that will ruin the connection to this PR.
If I want to reorder commits I usually do "interactive rebase", but in this particular case I think a non-interactive rebase is the best strategy:

1. You did your code changes on branch `bugfix/297-fix-picmg-led-get`
2. Now you create a new temporary local branch for magic cleanup:
    `git checkout -b tmpbranch origin/master` (that is, the new branch doesn't contain your code changes)
3. You do you magic cleanup and commit it to `tmpbranch`:
    `git add -u; git commit -s`
4. You switch to  `bugfix/297-fix-picmg-led-get`:
    `git checkout bugfix/297-fix-picmg-led-get`
5. You rebase this branch to your `tmpbranch`:
    `git rebase tmpbranch`
6. It will probably stop in the middle of the process complaining about conflicts.
    You edit the conflicting files (you can see them with `git status`), resolve the conflicts, and continue.
    Make sure there are no more conflict markers (`>>>`) in the conflicting files before you do this:
    `git add -u; git rebase --continue`
7. You forcefully push this branch to github:
    `git push -f`
8. Voila!

Reply:
As to conventions regarding magic, it is simple: don't use magic. Use macros, enums, or constants instead of dire numbers. When one uses magic numbers, it is always a hard guess whether this particular '32' is the same '32' as two lines below or are they semantically different.

That's reflected here:
https://github.com/ipmitool/ipmitool/wiki/Coding-Standards#magic-numbers

Reply:
This project needs a lot of this magic cleanup. That's an enormously huge task. So, to eat that elephant a bite at a time, I ask the contributors to try and eliminate the ancient magic wherever possible with their new contributions.

Reply:
Yes, replacing all magic numbers is a huge task. I initially didn't plan to touch the code that are not part of this issue. At the same time, I was trying to keep the same coding style at least within a same file. Otherwise, the final code would look weird as only a small portion of the code uses macros while the rest uses magic numbers.

I personally prefer to maintain the coding style consistency over fixing the numbers partially. That is, I wish having a dedicated issue/thread to fix the magic numbers. All other fixes/threads should keep the coding style the same. If the file was using macros, it should keep using macros. However, if the file was using raw numbers, it should keep using raw numbers.

Another argument is, for this project, a lot of magic numbers are from the specification. Anyone who works on the code shall  be familiar with the corresponding specifications. Therefore, s/he must know the meaning of those raw numbers. From this point of view, raw numbers are not really a big concern. On the other hand, it may be helpful to mention the chapter of the specification(s) the code is implementing.

Reply:
I respectfully but strongly disagree here. Magic numbers are NEVER allowed it a good maintainable code, regardless of them being by a specification or not. They ALWAYS MUST be represented as human-readable textual identifiers. There are tons of literature and online discussions on that topic that I don't want to repeat here.

Please, just follow the Coding Style currently in effect for this project and do not use magic numbers ever. I will strictly NOT accept any PRs with magic numbers.